### ISSUE TITLE
Publishing standalone app (C#) fails with C++/CLI dependency

### ISSUE THREAD

**Issue by chrisbardon** [dotnet/runtime#35269]
I'm trying to port a WPF app (C#) to core 3.1, but it has a dependency on a C++/CLI managed wrapper on an unmanaged DLL.  The dependency works, and I can build/run and publish as Framework Dependent, but I get a failure when I try to publish self contained.  The error that I get in VS 16.5.4 is:

> Publish has encountered an error.
> Publish has encountered an error. We were unable to determine the cause of the error. Check the output log for more details. 

And the .tmp file it generates tells me:

> 4/21/2020 8:58:07 PM
> System.AggregateException: One or more errors occurred. ---> System.Exception: Publish has encountered an error. We were unable to determine the cause of the error. Check the output log for more details. 
>    --- End of inner exception stack trace ---
> ---> (Inner Exception #0) System.Exception: Publish has encountered an error. We were unable to determine the cause of the error. Check the output log for more details. <---
> 
> System.Exception: Publish has encountered an error. We were unable to determine the cause of the error. Check the output log for more details. 

Which is not helpful...  The publish output however, had this buried in it:

> 3>Task "NETSdkError" skipped, due to false condition; ('$(Language)' == 'C++' and $(EnableComHosting) == 'true') was evaluated as ('C++' == 'C++' and  == 'true').
> 3>Using "NETSdkError" task from assembly "C:\Program Files\dotnet\sdk\3.1.201\Sdks\Microsoft.NET.Sdk\targets\..\tools\net472/Microsoft.NET.Build.Tasks.dll".
> 3>C++/CLI projects targeting .NET Core cannot use SelfContained=true.
> 3>Building with tools version "Current".

That implies that there's no way to build this as a self contained application, which was part of the motivation for going to .net core with this app in the first place (zero dependency deployment).  Is there a way to get this to work, or is it a bug?

**Comment by vitek-karas** [comment]
I think this is because we don't allow loading "components" as self-contained. Is that right @jkoritzinsky ?

**Comment by jkoritzinsky** [comment]
Building a self-contained C++/CLI application is not supported, but I remember some discussions with the SDK team about if C++/CLI dependencies of a self-contained C# app should be supported. @wli3 do you remember where we landed on that?

**Comment by wli3** [comment]
>  C++/CLI dependencies of a self-contained C# app should be supported.

That is true. From the error code message $(Language) is evaluated to C++. It might be caused by running _/t:publish_ in the folder. So all projects get published. Maybe try only publish the WPF project.

**Comment by CrapuleJack** [comment]
Just for the record, I too am having the same problem. I hope this is a feature .net 5 will bring.

For now I rely on the fact that my app directs the user to download and install the CLR at first run, wich is not ideal. A self-contained publish would be.

**Comment by FunksMaName** [comment]
I am experiencing this issue as well, and it would be great if Microsoft can just provide an official stance on the state of affairs, by clearly stating this  limitation on their documentation, here are a few places to start: 

Your **official** documentation page on publishing self contained apps for various runtimes
https://docs.microsoft.com/en-us/dotnet/core/deploying/deploy-with-cli

Blog entry touting the new C++/CLI project prowess which does not state this very important limitation
https://devblogs.microsoft.com/cppblog/porting-a-c-cli-project-to-net-core/ 

It's a very clumsy experience for our end user when we now have to make them download the Runtime prior to our app installation.

**Comment by FunksMaName** [comment]
@wli3 Just to remove any ambiguity, attempting to self contained publish just the project which takes a dependency on the C++/CLI project as you appear to suggest does not work either. This is what I get 

`C:\Program Files\dotnet\sdk\3.1.201\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(181,5): error NETSDK1121: C++/CLI projects targeting .NET Core cannot use SelfContained=true. [C:\Dev\yyy\src\yyy\yyy.Driver.Bridge.vcxproj]`

**Comment by CrapuleJack** [comment]
An other option would be to provide a .netcore runtime merge module I could include in my setups .msi, but I cant find those if they exists, anyone knows about that?

This would be an alternative but the ideal would still be self contained deploy.

**Comment by FunksMaName** [comment]
@CrapuleJack my work around was to a WIX Bundle project to install the dependency, if you are using wix then here is a reference 

```xml<?xml version="1.0" encoding="UTF-8"?>
<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi" 
     xmlns:bal="http://schemas.microsoft.com/wix/BalExtension">
	<Bundle 
    Name="YYYY Driver Installer (Bundle)" 
    Version="1.0.0.0" 
    Manufacturer="YYYY" 
    UpgradeCode="F6396F05-BEDF-4B14-AD41-57C91CBC8772"
    IconSourceFile="app.ico">
    <BootstrapperApplicationRef Id="WixStandardBootstrapperApplication.RtfLicense">
      <bal:WixStandardBootstrapperApplication
        LicenseFile="License.rtf"
        LogoFile="Logo.png" />
    </BootstrapperApplicationRef>
    <Chain>
	  <!-- Dot Net Hosting Runtime exe
	  (the command -q seems to make it install silently but not documented anywhere)  -->
      <ExePackage SourceFile="dotnet-hosting-3.1.5-win.exe"
                  InstallCommand="-q"
                  UninstallCommand="-q">
        <ExitCode Value="0" Behavior="success" />
      </ExePackage>
      <RollbackBoundary />
      <!-- Your Product MSI -->
      <MsiPackage 
        SourceFile="YourInstallationMSI.msi"
        DisplayInternalUI="no" /> 
    </Chain>
	</Bundle>
</Wix>
```
Some resources around WIX Boostrapping
https://stackoverflow.com/questions/47399366/wix-how-do-i-launch-another-msi-or-exe-inside-my-installer-c
https://www.firegiant.com/wix/tutorial/net-and-net/bootstrapping/

**Comment by CrapuleJack** [comment]
@FunksMaName Thanks a lot for that, I will take a look at that for sure
